### ISSUE TITLE
all zero in input_output_TF_0.9quantile.GCN.string.npy

### ISSUE THREAD

**Issue by trista1115** [xixi-cathy/regX#3]
Hello, I used the T2D data and ran the code according to the instructions, but the output file 'input_output_TF_0.9quantile.GCN.string.npy' contains all zeros. Checking upstream steps, I found that all differences were zero during the perturbation process. I'm not sure exactly where the problem lies. Do you have any suggestions? below is my part of test code and output information :

for i in range(repeats):
    model = Net(num_genes=expr_test.shape[1], num_peaks=num_peaks, num_tf=X_test.shape[1], cut=cut).to(device)
    model.eval()
    model.load_state_dict(torch.load('./predict_status/model.GCN.string.'+str(i)+'.pt', map_location=device))
    np.random.seed(123) 

    _, _, orig = model(X_test)
    orig = F.softmax(orig)
    print(f"Original model output (rep {i}): {orig[:5]}")

    for j in range(X_test.shape[1]):
        X_new = torch.clone(X_test).detach()
        tf_mat = cell_by_tf[indices_test, j] + np.quantile(cell_by_tf[:, j], 0.9)
        tf_mat = np.repeat(tf_mat[:, np.newaxis], X_test.shape[2], axis=1)
        peak_mat = cell_by_peak[indices_test, :]
        w = W[j, :]
        w = np.repeat(w[np.newaxis, :], X_test.shape[0], axis=0)
        new = tf_mat*peak_mat*w
        if np.all(new == 0):
            print(f"Warning: new matrix all zeros for TF {j}")
        X_new[:, j, :] = torch.from_numpy(new).float().to(device)
        _, _, perturb = model(X_new)
        perturb = F.softmax(perturb)
        difference = (perturb - orig).detach().cpu().numpy()
        if np.all(difference == 0):
            print(f"Warning: No difference for TF {j} in repeat {i}")
        results[i, :, j, :] = (perturb-orig).detach().cpu().numpy()
        results_orig[i, :, j, :] = orig.clone().detach().cpu().numpy()
        results_perturb[i, :, j, :] = perturb.detach().cpu().numpy()

`np.save('./predict_status/interpret_test/input_output_TF_0.9quantile.GCN.string.npy', results)`

Warning: No difference for TF 87 in repeat 0
Warning: No difference for TF 88 in repeat 0
Warning: new matrix all zeros for TF 89
Warning: No difference for TF 89 in repeat 0
Warning: No difference for TF 90 in repeat 0
Warning: No difference for TF 91 in repeat 0
Warning: No difference for TF 92 in repeat 0
Warning: No difference for TF 93 in repeat 0
.....
Original model output (rep 6): tensor([[1.3265e-06, 3.3257e-04, 9.9967e-01],
        [6.6395e-05, 9.9959e-01, 3.4821e-04],
        [9.9998e-01, 6.7583e-06, 1.2186e-05],
        [4.7350e-02, 9.4734e-01, 5.3118e-03],
        [7.3250e-04, 5.8701e-02, 9.4057e-01]], device='cuda:3',
       grad_fn=<SliceBackward0>)
Warning: new matrix all zeros for TF 89
Warning: new matrix all zeros for TF 167
Warning: new matrix all zeros for TF 195
Original model output (rep 7): tensor([[0.3120, 0.3231, 0.3649],
        [0.3120, 0.3231, 0.3649],
        [0.3120, 0.3231, 0.3649],
        [0.3120, 0.3231, 0.3649],
        [0.3120, 0.3231, 0.3649]], device='cuda:3', grad_fn=<SliceBackward0>)
Warning: No difference for TF 0 in repeat 7
Warning: No difference for TF 1 in repeat 7

**Comment by xixi-cathy** [comment]
Hi,

Thanks for your interest in our work. Did you use the models we provided for perturbation? Or did you use the models trained by yourself?

Xi

**Comment by trista1115** [comment]
> Hi,
> 
> Thanks for your interest in our work. Did you use the models we provided for perturbation? Or did you use the models trained by yourself?
> 
> Xi

I applied the perturbation model from the T2D step-by-step workflow.
In addition, I obtained 1,399 aggregated cells instead of the expected 1,394 cells using the 0.preprocess_T2D.R.ipynb script.

**Comment by xixi-cathy** [comment]
This is weird, cause I have not come across such an issue. I mean, did you download the model parameters from Zenodo that we provided, or train the model by yourself?
A varying number of aggregated cells is possible, since the aggregation algorithm was based on UMAP, which might be different when you run "RunUMAP" each time. But this should not affect much, and not be the reason for this.

**Comment by trista1115** [comment]
> This is weird, cause I have not come across such an issue. I mean, did you download the model parameters from Zenodo that we provided, or train the model by yourself? A varying number of aggregated cells is possible, since the aggregation algorithm was based on UMAP, which might be different when you run "RunUMAP" each time. But this should not affect much, and not be the reason for this.

Thanks for your quick reply. I did not use the model parameters from Zenodo but instead downloaded them from GitHub.

Additionally, I successfully ran the demo example (mouse) using your GitHub code (regX.1.0). Now, I am re-running the model with the T2D data using the scripts from ./regX.1.0/custom_code, and the process is still ongoing.

By the way, could you share the Zenodo link for your model parameters?

**Comment by xixi-cathy** [comment]
The download link in GitHub is directed to Zenodo (https://doi.org/10.5281/zenodo.11608076), so we should be talking about the same thing. 
I just realized that if the aggregated samples you used do not match the ones used to train models, problems might appear. In other words, the model parameters we provided were trained using the 1,394 aggregated cells. But if you use the 1,399 newly aggregated cells and models trained on 1,394 cells for downstream analysis, things might go wrong. I'm not sure whether this is the reason.
Looking forward to your feedback on using the custom code.

**Comment by trista1115** [comment]
> The download link in GitHub is directed to Zenodo (https://doi.org/10.5281/zenodo.11608076), so we should be talking about the same thing. I just realized that if the aggregated samples you used do not match the ones used to train models, problems might appear. In other words, the model parameters we provided were trained using the 1,394 aggregated cells. But if you use the 1,399 newly aggregated cells and models trained on 1,394 cells for downstream analysis, things might go wrong. I'm not sure whether this is the reason. Looking forward to your feedback on using the custom code.

Sorry about confusion earlier. I now understand you had mentioned this previously. To clarify, I trained the model using 1,399 cells by myself (rather than your provided parameters trained on 1,394 cells). Here's what happened:

1.Initially, I used your ./regX.1.0/example_code to train on T2D data (1,399 cells) by myself. but the output figures lacked edges. Upon checking upstream outputs, I found zero differences during perturbation:
`differences = (perturb - orig).detach().cpu().numpy()`
And I am still dont know why this issue happened ?

2.Currently, I'm running the T2D data through ./regX.1.0/custom_code (ongoing). This approach was chosen because I  successfully ran the demo (mouse) example.

Thanks for your time~
